virtwho/configure.py

```diff
@@ -1,4 +1,124 @@
-"""
-Used to manage virt-who configure file
-Add/Delete/Update options for /etc/virt-who.conf, /etc/virt-who.d/xxx.conf
-"""
+import os
+from virtwho.settings import Configure
+from virtwho.settings import config
+from virtwho.settings import TEMP_DIR
+from virtwho.ssh import SSHConnect
+from virtwho.logger import getLogger
+
+logger = getLogger(__name__)
+
+
+class VirtwhoHypervisorConfig:
+    """Able to create and manage /etc/virt-who.d/xxx.conf file when
+    call this class"""
+    def __init__(self, mode='esx', register_type='rhsm'):
+        """Create virt-who configuration file with basic options.
+         All data come from virtwho.ini file.
+        :param mode: The hypervisor mode.
+            (esx, xen, hyperv, rhevm, libvirt, kubevirt, local)
+        :param register_type: The subscription server. (rhsm, satellite)
+        """
+        self.mode = mode
+        self.register_type = register_type
+        self.remote_ssh = virtwho_ssh_connect(self.mode)
+        self.local_file = os.path.join(TEMP_DIR, f'{self.mode}.conf')
+        self.remote_file = f'/etc/virt-who.d/{self.mode}.conf'
+        self.section = f'virtwho-{self.mode}'
+        self.cfg = Configure(self.local_file, self.remote_ssh, self.remote_file)
+        if self.mode == 'local':
+            self.update('type', 'libvirt')
+        else:
+            if self.register_type == 'rhsm':
+                register = config.rhsm
+            elif self.register_type == 'satellite':
+                register = config.satellite
+            if self.mode == 'esx':
+                hypervisor = config.vcenter
+            elif self.mode == 'xen':
+                hypervisor = config.xen
+            elif self.mode == 'hyperv':
+                hypervisor = config.hyperv
+            elif self.mode == 'rhevm':
+                hypervisor = config.rhevm
+            elif self.mode == 'libvirt':
+                hypervisor = config.libvirt
+            elif self.mode == 'kubevirt':
+                hypervisor = config.kubevirt
+            self.update('type', self.mode)
+            self.update('hypervisor_id', 'hostname')
+            self.update('owner', register.default_org)
+            if self.mode == 'kubevirt':
+                self.update('kubeconfig', hypervisor.config_file)
+            else:
+                self.update('server', hypervisor.server)
+                self.update('username', hypervisor.username)
+                self.update('password', hypervisor.password)
```

Comment:
1. if the mode is not used by other functions(update/delete/destroy), it's not necessary to define `self.mode`, mode is everything.

2. we may use two separated functions to return the hypervisor and register handle just like we get the virt-who host ssh handle.
```
def get_register_handler(register_type):
    register = config.rhsm
    if self.register_type == 'satellite':
        register = config.satellite
    return register

def get_hypervisor_handler(mode):
    hypervisor = config.vcenter
    if self.mode == 'xen':
        hypervisor = config.xen
    if self.mode == 'hyperv':
        hypervisor = config.hyperv
    if self.mode == 'rhevm':
        hypervisor = config.rhevm
    if self.mode == 'libvirt':
        hypervisor = config.libvirt
    if self.mode == 'kubevirt':
        hypervisor = config.kubevirt
    if self.mode == 'local':
        hypervisor = config.local
    return hypervisor
```

3. so the __init__ function is easy like:
```
    def __init__(self, mode='esx', register_type='rhsm'):
        self.register = get_register_handler(register_type)
        self.hypervisor = get_hypervisor_handler(mode)
        self.remote_ssh = virtwho_host_connect(mode)
        self.local_file = os.path.join(TEMP_DIR, f'{mode}.conf')
        self.remote_file = f'/etc/virt-who.d/{mode}.conf'
        self.cfg = Configure(self.local_file, self.remote_ssh, self.remote_file)

        self.section = f'virtwho-{mode}'
        self.update('type', mode)
        if mode is not 'local'
            self.update('hypervisor_id', 'hostname')
            self.update('owner', self.register.default_org)
        if mode is 'kubevirt':
            self.update('kubeconfig', hypervisor.config_file)
        if mode in ('esx','xen','hyperv','rhemv','libvirt'):
            self.update('server', self.hypervisor.server)
            self.update('username', self.hypervisor.username)
            self.update('password', self.hypervisor.password)
```

Reply:
Done to split into get_register_handler() and get_hypervisor_handler() functions, which makes the code more clear and easier to understand.